driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,119 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+class BatchCursorFlux<T> implements Publisher<T> {
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private volatile BatchCursor<T> batchCursor;
+    private FluxSink<T> sink;
+
+    BatchCursorFlux(final BatchCursorPublisher<T> batchCursorPublisher) {
+        this.batchCursorPublisher = batchCursorPublisher;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        Flux.<T>create(sink -> {
+            this.sink = sink;
+            sink.onRequest(demand -> {
+                if (calculateDemand(demand) > 0 && inProgress.compareAndSet(false, true)) {
+                    if (batchCursor == null) {
+                        int batchSize = calculateBatchSize(sink.requestedFromDownstream());
+                        batchCursorPublisher.batchCursor(batchSize).subscribe(bc -> {
+                            batchCursor = bc;
+                            inProgress.set(false);
+                            recurseCursor();
+                        }, sink::error);
+                    } else {
+                        inProgress.set(false);
+                        recurseCursor();
+                    }
+                }
+            });
+            sink.onCancel(this::closeCursor);
+            sink.onDispose(this::closeCursor);
+        }, FluxSink.OverflowStrategy.BUFFER)
+        .subscribe(subscriber);
+    }
+
+    private void closeCursor() {
+        if (batchCursor != null) {
+            batchCursor.close();
+        }
+    }
+
+    private void recurseCursor(){
+        if (!sink.isCancelled() && sink.requestedFromDownstream() > 0 && inProgress.compareAndSet(false, true)) {
+            if (batchCursor.isClosed()) {
+                sink.complete();
+            } else {
+                batchCursor.setBatchSize(calculateBatchSize(sink.requestedFromDownstream()));
+                Mono.from(batchCursor.next())
+                        .doOnCancel(this::closeCursor)
+                        .doOnError((e) -> {
+                            closeCursor();
+                            sink.error(e);
```

Comment:
If for some reason `closeCursor` throws, we are not going to call `sink.error`. Is there a good reason to not protect ourselves from such a scenario by using `try { close } finally { sink.error }`?

Reply:
Its probably overly defensive as if it errors then something else is wrong.  It should be idempotent.

Reply:
Done

Reply:
Thank you! I would agree that it's overly defensive if we were calling `System.exit` or `Runtime.halt` methods as a result of failed `closeCursor`. As long as it's not the case, the code needs to do its best to try and follow the contracts.